### ISSUE TITLE
CALL_FUNCTION_VAR opcode

### ISSUE THREAD

**Issue by Wizr** [odoo/odoo#13166]
Impacted versions:
9.0

Steps to reproduce:
evaluate this expression in a qweb template: `(lambda a, b: 1+2)(*(1, 2))`

Current behavior:
`raise ValueError("forbidden opcode(s) in %r" % expr)`

Expected behavior:
opcodes `CALL_FUNCTION_VAR, CALL_FUNCTION_KW, CALL_FUNCTION_VAR_KW` should be allowed

Video/Screenshot link (optional):

**Comment by dreispt** [comment]
Why?

**Comment by Wizr** [comment]
@dreispt Unpacking argument list(the asterisk) in the function call require opcode `CALL_FUNCTION_VAR `. I don't think there is any reason to disable this feature.

**Comment by dreispt** [comment]
I think the problem is the lambda use, not the unpacking.

**Comment by Wizr** [comment]
I'm sure that lambda use is allowed, such as `(lambda a, b: 1+2)(1, 2)`.

**Comment by odony** [comment]
You're right, I don't see any reason to block `CALL_FUNCTION_VAR`, `CALL_FUNCTION_KW` and `CALL_FUNCTION_VAR_KW` given that we allow `CALL_FUNCTION`.
The workaround is trivial however: unpack the arguments before calling the function. Because of that, I'd rather fix it in master only, to avoid the noise in stable series.

**Comment by Wizr** [comment]
@odony I don't think we can unpack the argument and then pass it to the function in one expression.

**Comment by odony** [comment]
@Wizr Why not, this should work, right? ``f(*args) -> f(args[0], args[1], ...)``

```python
>>> args = (1,2)
>>> def a():
...     (lambda a, b: a+b)(*args)
... 
>>> dis.dis(a)
  2           0 LOAD_CONST               1 (<code object <lambda> at 0x7f886c661b30, file "<stdin>", line 2>)
              3 MAKE_FUNCTION            0
              6 LOAD_GLOBAL              0 (args)
              9 CALL_FUNCTION_VAR        0
             12 POP_TOP             
             13 LOAD_CONST               0 (None)
             16 RETURN_VALUE        
>>> def b():
...     # no need for CALL_FUNCTION_VAR
...     (lambda a, b: a+b)(args[0], args[1])
... 
>>> dis.dis(b)
  2           0 LOAD_CONST               1 (<code object <lambda> at 0x7f886c6b9b30, file "<stdin>", line 2>)
              3 MAKE_FUNCTION            0
              6 LOAD_GLOBAL              0 (args)
              9 LOAD_CONST               2 (0)
             12 BINARY_SUBSCR       
             13 LOAD_GLOBAL              0 (args)
             16 LOAD_CONST               3 (1)
             19 BINARY_SUBSCR       
             20 CALL_FUNCTION            2
             23 POP_TOP             
             24 LOAD_CONST               0 (None)
             27 RETURN_VALUE        
>>> 
```

And if you are using QWeb specifically, you can split expressions using intermediary variable, for example with ``<t t-set="var" t-value="expr"/>``

**Comment by Wizr** [comment]
@odony I took the argument as an arbitrary length list, but I can't think of any situation where we need it. So you're right.

When talking about `safe_eval`, I noticed that `*_CLOSURE` (by embed lambda in lambda) and some of the builtin functions (like `reversed`) is also not allowed.  Is it safety concern or by design to avoid complexity?

**Comment by dmo-odoo** [comment]
@odony Due to the dle effect, we encountered an issue related to this with @stefanorigano just a few minutes ago. I discussed with @KangOl who was also okay to allow those opcodes. Could we do it in 10.0 ?

**Comment by odony** [comment]
This has been fixed in 11.0 as planned, so we can close it.

@Wizr: `reversed` is mostly missing because we never had the use for it, and `*_CLOSURE` is for simplicity reasons. This feature was designed for evaluating rather simple expressions, not developing complex pieces of logic, so we never felt the need for those :-)
